File: imba/src/utils/scoreSearch.js
/**
 * Compute search relevance score for a string.
 * 
 * @param {string} text
 * @param {string} query
 * @returns {number} score
 */
export function scoreSearch(text, query) {
  if (!text || !query) {
    return 0;
  }

  let searchText = text.toLowerCase();
  let score = 0;
  let p = 0; // Position within the `text`
  let consecutiveBonus = 2; // Bonus for consecutive matches

  // Here is the main magic
  // Look through each character of the queryText string, stopping at the end(s)...
  let queryLength = query.length;
  for (let i = 0; i < queryLength; i += 1) {
    // Figure out if the current letter is found in the rest of the `text`.
    let index = searchText.indexOf(query[i], p);
    // If not, continue to the next character.
    if (index < 0) {
      // If the character is not found after p, check if it is found before p
      // Because we are jumping over the characters, we need to check if the character is found before p
      index = searchText.indexOf(query[i]);
      if (index < 0) continue;
    }
    //  If it is, add to the score...
    score += 1;
    // If the character is found in the next two chars, give it a bonus for being consecutive
    if (index - p < 2) {
      score += consecutiveBonus;
      // if the char is next, multiply the bonus by 2
      if (index === p) consecutiveBonus *= 2;
      // Otherwise, don't reset the bonus, maybe user missed a char
    } else {
      consecutiveBonus = 2;
    }
    //  ... and skip the position within `text` forward.
    p = index + 1;
  }

  // bonus for exact match & similar length
  if (
    searchText.includes(query) &&
    Math.abs(searchText.length - query.length) < 2
  ) {
    score += 2;
  }

  // filter by relative threshold
  if (score >= query.length * 2) {
    return score;
  }
  return 0;
}

File: django/bolls/utils/score_search.py
def score_search(text: str, query: str) -> int:
    if not text or not query:
        return 0

    search_text = text.lower()
    score = 0
    p = 0  # Position within the `text`
    consecutive_bonus = 2  # Bonus for consecutive matches

    query_length = len(query)
    for i in range(query_length):
        index = search_text.find(query[i], p)
        if index < 0:
            index = search_text.find(query[i])
            if index < 0:
                continue

        score += 1
        if index - p < 2:
            score += consecutive_bonus
            if index == p:
                consecutive_bonus *= 2
        else:
            consecutive_bonus = 2

        p = index + 1

    if query in search_text:
        score += 2

    if score >= len(query):
        return score

    return 0


# a few testcases
# print(score_search('sing', 'ED SH ing'))
# print(score_search("I Don't care", 'DON‘T CARE CHERYL'))
# print(score_search("Cheryl", 'DON‘T CARE CHERYL'))
# print(score_search('Good Enough', 'GOOD EN'))
# print(score_search('God Save the Queen', 'GOOD EN'))
# print(score_search('Del Shannon', 'ed shran'))
# print(score_search('Ed Sheeran', 'ed shran'))
